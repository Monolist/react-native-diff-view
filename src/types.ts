import { IHunk, IChange } from 'gitdiff-parser';
import { TextStyle, ViewStyle } from 'react-native';

export type DiffType = 'add' | 'delete' | 'modify' | 'rename' | 'copy';

export type Side = 'old' | 'new';

export type ViewType = 'unified' | 'split';

export type GutterType = 'default' | 'none' | 'anchor';

export type Widgets = Record<string, JSX.Element>;

export type Events = Record<string, (arg: any) => any>;

export type DiffChildren = (hunks: IHunk[]) => JSX.Element | JSX.Element[];

export interface IGutterOptions {
  change: IChange;
  side: Side;
  inHoverState: boolean;
  textStyle?: TextStyle;
  renderDefault: (style?: TextStyle, textStyle?: ViewStyle) => JSX.Element;
}
