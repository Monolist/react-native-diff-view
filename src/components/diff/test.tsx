import React from 'react';
import renderer from 'react-test-renderer';
import { IHunk, DiffChildren, Widgets } from '../../types';
import { basicHunk } from '../../fixtures';
import { Decoration, getChangeKey } from '../..';
import Diff from '.';

const TestDiff = ({ children }: { children: DiffChildren }): JSX.Element => (
  <Diff diffType={'modify'} hunks={[basicHunk]}>
    {children}
  </Diff>
);

describe('components/diff', () => {
  const renderRawHunks = (hunks: IHunk[]): JSX.Element => <div>{JSON.stringify(hunks)}</div>;

  describe('base', () => {
    it('renders correctly', () => {
      expect(renderer.create(<TestDiff>{renderRawHunks}</TestDiff>).toJSON()).toMatchSnapshot();
    });
  });

  describe('w/ Decoration', () => {
    const renderDecoration = (): JSX.Element => (
      <Decoration hideGutter={false} monotonous={false}>
        <div>xxx</div>
      </Decoration>
    );

    it('renders correctly', () => {
      expect(renderer.create(<TestDiff>{renderDecoration}</TestDiff>).toJSON()).toMatchSnapshot();
    });
  });

  describe('w/ Widget', () => {
    const getWidgets = (hunks: IHunk[]): Widgets => {
      const changes = hunks.reduce((result, { changes }) => [...result, ...changes], []);
      const longLines = changes.filter(({ content }) => content.length > 20);

      return longLines.reduce((widgets, change) => {
        const changeKey = getChangeKey(change);

        return {
          ...widgets,
          [changeKey]: <span className="error">Line too long</span>,
        };
      }, {});
    };

    it('renders correctly', () => {
      const hunks = [basicHunk];

      expect(
        renderer
          .create(
            <Diff hunks={hunks} widgets={getWidgets(hunks)} diffType="modify">
              {renderRawHunks}
            </Diff>,
          )
          .toJSON(),
      ).toMatchSnapshot();
    });
  });
});
