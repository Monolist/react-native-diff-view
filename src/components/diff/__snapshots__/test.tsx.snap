// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/diff base renders correctly 1`] = `
<View
  style={Object {}}
>
  <View>
    <View />
    <View />
    <View />
    <View />
  </View>
  <div>
    [{"content":"@@ -21,3 +21,3 @@ describe('basic test', () =&gt; {","oldStart":21,"newStart":21,"oldLines":3,"newLines":3,"changes":[{"content":"    it('App renders correctly', () =&gt; {","type":"normal","isNormal":true,"oldLineNumber":21,"newLineNumber":21},{"content":"        expect(renderer.create(&lt;App diffText={'deff'} /&gt;).toJSON()).toMatchSnapshot();","type":"delete","isDelete":true,"lineNumber":22},{"content":"        expect(renderer.create(&lt;App diffText={'diff'} /&gt;).toJSON()).toMatchSnapshot();","type":"insert","isInsert":true,"lineNumber":22},{"content":"    });","type":"normal","isNormal":true,"oldLineNumber":23,"newLineNumber":23}],"isPlain":false}]
  </div>
</View>
`;

exports[`components/diff base unified Diff 1`] = `
<View
  style={Object {}}
>
  <View>
    <View />
    <View />
    <View />
  </View>
  <div>
    [{"content":"@@ -21,3 +21,3 @@ describe('basic test', () =&gt; {","oldStart":21,"newStart":21,"oldLines":3,"newLines":3,"changes":[{"content":"    it('App renders correctly', () =&gt; {","type":"normal","isNormal":true,"oldLineNumber":21,"newLineNumber":21},{"content":"        expect(renderer.create(&lt;App diffText={'deff'} /&gt;).toJSON()).toMatchSnapshot();","type":"delete","isDelete":true,"lineNumber":22},{"content":"        expect(renderer.create(&lt;App diffText={'diff'} /&gt;).toJSON()).toMatchSnapshot();","type":"insert","isInsert":true,"lineNumber":22},{"content":"    });","type":"normal","isNormal":true,"oldLineNumber":23,"newLineNumber":23}],"isPlain":false}]
  </div>
</View>
`;

exports[`components/diff w/ Decoration renders correctly 1`] = `
<View
  style={Object {}}
>
  <View>
    <View />
    <View />
    <View />
    <View />
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "flex": 1,
        "flexDirection": "row",
        "lineHeight": 1.5,
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
        }
      }
    >
      <div>
        xxx
      </div>
    </View>
  </View>
</View>
`;

exports[`components/diff w/ Decoration unified Diff with Decoration 1`] = `
<View
  style={Object {}}
>
  <View>
    <View />
    <View />
    <View />
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "flex": 1,
        "flexDirection": "row",
        "lineHeight": 1.5,
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
        }
      }
    >
      <div>
        xxx
      </div>
    </View>
  </View>
</View>
`;

exports[`components/diff w/ Widget split widget 1`] = `
<View
  style={Object {}}
>
  <View>
    <View />
    <View />
    <View />
    <View />
  </View>
  <div>
    [{"content":"@@ -21,3 +21,3 @@ describe('basic test', () =&gt; {","oldStart":21,"newStart":21,"oldLines":3,"newLines":3,"changes":[{"content":"    it('App renders correctly', () =&gt; {","type":"normal","isNormal":true,"oldLineNumber":21,"newLineNumber":21},{"content":"        expect(renderer.create(&lt;App diffText={'deff'} /&gt;).toJSON()).toMatchSnapshot();","type":"delete","isDelete":true,"lineNumber":22},{"content":"        expect(renderer.create(&lt;App diffText={'diff'} /&gt;).toJSON()).toMatchSnapshot();","type":"insert","isInsert":true,"lineNumber":22},{"content":"    });","type":"normal","isNormal":true,"oldLineNumber":23,"newLineNumber":23}],"isPlain":false}]
  </div>
</View>
`;

exports[`components/diff w/ Widget unified widget 1`] = `
<View
  style={Object {}}
>
  <View>
    <View />
    <View />
    <View />
  </View>
  <div>
    [{"content":"@@ -21,3 +21,3 @@ describe('basic test', () =&gt; {","oldStart":21,"newStart":21,"oldLines":3,"newLines":3,"changes":[{"content":"    it('App renders correctly', () =&gt; {","type":"normal","isNormal":true,"oldLineNumber":21,"newLineNumber":21},{"content":"        expect(renderer.create(&lt;App diffText={'deff'} /&gt;).toJSON()).toMatchSnapshot();","type":"delete","isDelete":true,"lineNumber":22},{"content":"        expect(renderer.create(&lt;App diffText={'diff'} /&gt;).toJSON()).toMatchSnapshot();","type":"insert","isInsert":true,"lineNumber":22},{"content":"    });","type":"normal","isNormal":true,"oldLineNumber":23,"newLineNumber":23}],"isPlain":false}]
  </div>
</View>
`;
